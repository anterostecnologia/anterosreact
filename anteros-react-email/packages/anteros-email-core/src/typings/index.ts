import { IPage } from '@core/blocks';
import { ReactElement } from 'react';

export interface IBlock<T extends IBlockData = IBlockData> {
  name: string;
  type: string;
  create: (payload?: RecursivePartial<T>) => T;
  validParentType: string[];
  render?: (
    data: T,
    idx: string | null,
    mode: 'testing' | 'production',
    context?: IPage,
    dataSource?: { [key: string]: any; }
  ) => IBlockData | ReactElement | null;
}

export interface IBlockData<
  K extends { [key: string]: any; } = any,
  T extends { [key: string]: any; } = any
  > {
  title?: string;
  type: string;
  data: {
    value: T;
    hidden?: boolean | string;
  };
  attributes: K & { 'css-class'?: string; };
  children: IBlockData[];
}

export interface create<T extends any = any> {
  (payload?: RecursivePartial<T>): T;
}

export type RecursivePartial<T> = {
  [P in keyof T]?: T[P] extends (infer U)[]
  ? RecursivePartial<U>[]
  : T[P] extends object
  ? RecursivePartial<T[P]>
  : T[P];
};
