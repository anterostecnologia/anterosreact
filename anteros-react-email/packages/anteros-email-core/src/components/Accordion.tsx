
import { omit } from 'lodash';
import { BasicType } from '@core/constants';
import { RecursivePartial } from '@core/typings';
import React from 'react';
import { IAccordion } from '@core/blocks';
import MjmlBlock, { MjmlBlockProps } from '@core/components/MjmlBlock';

export type AccordionProps = RecursivePartial<IAccordion['data']> &
  RecursivePartial<IAccordion['attributes']> & {
    children?: MjmlBlockProps<IAccordion>['children'];
  };

export function Accordion(props: AccordionProps) {
  return (
    <MjmlBlock
      attributes={omit(props, ['data', 'children', 'value'])}
      value={props.value}
      type={BasicType.ACCORDION}
    >
      {props.children}
    </MjmlBlock>
  );
}
