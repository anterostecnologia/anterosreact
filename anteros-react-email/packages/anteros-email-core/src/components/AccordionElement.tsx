
import { omit } from 'lodash';
import { BasicType } from '@core/constants';
import { RecursivePartial } from '@core/typings';
import React from 'react';
import { IAccordionElement } from '@core/blocks';
import MjmlBlock, { MjmlBlockProps } from '@core/components/MjmlBlock';

export type AccordionElementProps = RecursivePartial<IAccordionElement['data']> &
  RecursivePartial<IAccordionElement['attributes']> & {
    children?: MjmlBlockProps<IAccordionElement>['children'];
  };

export function AccordionElement(props: AccordionElementProps) {
  return (
    <MjmlBlock
      attributes={omit(props, ['data', 'children', 'value'])}
      value={props.value}
      type={BasicType.ACCORDION_ELEMENT}
    >
      {props.children}
    </MjmlBlock>
  );
}
