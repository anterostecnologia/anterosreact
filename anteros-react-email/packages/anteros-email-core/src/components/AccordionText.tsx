
import { omit } from 'lodash';
import { BasicType } from '@core/constants';
import { RecursivePartial } from '@core/typings';
import React from 'react';
import { IAccordionText } from '@core/blocks';
import MjmlBlock, { MjmlBlockProps } from '@core/components/MjmlBlock';

export type AccordionTextProps = RecursivePartial<IAccordionText['data']> &
  RecursivePartial<IAccordionText['attributes']> & {
    children?: MjmlBlockProps<IAccordionText>['children'];
  };

export function AccordionText(props: AccordionTextProps) {
  return (
    <MjmlBlock
      attributes={omit(props, ['data', 'children', 'value'])}
      value={props.value}
      type={BasicType.ACCORDION_TEXT}
    >
      {props.children}
    </MjmlBlock>
  );
}
