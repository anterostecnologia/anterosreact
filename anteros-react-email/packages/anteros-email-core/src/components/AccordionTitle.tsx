
import { omit } from 'lodash';
import { BasicType } from '@core/constants';
import { RecursivePartial } from '@core/typings';
import React from 'react';
import { IAccordionTitle } from '@core/blocks';
import MjmlBlock, { MjmlBlockProps } from '@core/components/MjmlBlock';

export type AccordionTitleProps = RecursivePartial<IAccordionTitle['data']> &
  RecursivePartial<IAccordionTitle['attributes']> & {
    children?: MjmlBlockProps<IAccordionTitle>['children'];
  };

export function AccordionTitle(props: AccordionTitleProps) {
  return (
    <MjmlBlock
      attributes={omit(props, ['data', 'children', 'value'])}
      value={props.value}
      type={BasicType.ACCORDION_TITLE}
    >
      {props.children}
    </MjmlBlock>
  );
}
