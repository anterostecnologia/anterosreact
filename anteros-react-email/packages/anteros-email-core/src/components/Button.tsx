
import { omit } from 'lodash';
import { BasicType } from '@core/constants';
import { RecursivePartial } from '@core/typings';
import React from 'react';
import { IButton } from '@core/blocks';
import MjmlBlock, { MjmlBlockProps } from '@core/components/MjmlBlock';

export type ButtonProps = RecursivePartial<IButton['data']> &
  RecursivePartial<IButton['attributes']> & {
    children?: MjmlBlockProps<IButton>['children'];
  };

export function Button(props: ButtonProps) {
  return (
    <MjmlBlock
      attributes={omit(props, ['data', 'children', 'value'])}
      value={props.value}
      type={BasicType.BUTTON}
    >
      {props.children}
    </MjmlBlock>
  );
}
