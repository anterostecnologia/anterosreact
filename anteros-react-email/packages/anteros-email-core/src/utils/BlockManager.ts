import { IBlock, IBlockData } from '@core/typings';
import { standardBlocks, advancedBlocks } from '@core/blocks';

export class BlockManager {
  private static blocksMap = { ...standardBlocks, ...advancedBlocks };
  private static autoCompletePath: { [key: string]: Array<string[]>; } = {};

  private static setAutoCompletePath() {
    const paths: { [key: string]: Array<string[]>; } = {};

    const renderFullPath = (
      type: string,
      pathObj: Array<string[]>,
      prevPaths: string[]
    ): any => {
      const block = this.getBlockByType(type);
      if (!block) {
        throw new Error(`Can you register ${type} block`);
      }
      const currentPaths = [...prevPaths, type];
      if (block.validParentType.length === 0) {
        pathObj.push(currentPaths);
      }
      return block.validParentType.map((item) => {
        return renderFullPath(item, pathObj, currentPaths);
      });
    };

    Object.values(this.blocksMap).forEach((item) => {
      paths[item.type] = [];
      renderFullPath(item.type, paths[item.type], []);
    });
    return paths;
  }

  public static getBlocks(): Array<IBlock> {
    return Object.values(this.blocksMap);
  }

  public static registerBlocks(blocksMap: { [key: string]: IBlock; }) {
    this.blocksMap = {
      ...this.blocksMap,
      ...blocksMap,
    };
    this.autoCompletePath = this.setAutoCompletePath();
  }

  public static getBlockByType<T extends IBlockData>(
    type: string
  ): IBlock<T> | undefined {
    const map = this.getBlocksByType([type]);
    return map[0] as any;
  }

  public static getBlocksByType(
    types: Array<string>
  ): Array<IBlock | undefined> {
    return types.map((item) => {
      const block = Object.values(this.blocksMap).find((child) => {
        return child.type === item;
      });

      return block;
    });
  }

  public static getAutoCompleteFullPath() {
    if (Object.keys(this.autoCompletePath).length === 0) {
      this.autoCompletePath = this.setAutoCompletePath();
    }
    return this.autoCompletePath;
  }

  static getAutoCompletePath(
    type: string,
    targetType: string
  ): Array<string> | null {
    const block = this.getBlockByType(type);
    if (!block) {
      throw new Error(`Can you register ${type} block`);
    }
    if (block.validParentType.includes(targetType)) {
      return [];
    }
    const paths = this.getAutoCompleteFullPath()[type as any].find((item) =>
      item.filter((_, index) => index !== 0).includes(targetType)
    );

    if (!paths) return null;
    const findIndex = paths.findIndex((item) => item === targetType);
    return paths.slice(1, findIndex);
  }
}
