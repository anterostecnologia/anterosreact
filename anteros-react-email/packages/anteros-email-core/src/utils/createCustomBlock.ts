import { IPage } from '@core/blocks';
import { IBlock, IBlockData } from '@core/typings';
import { ReactElement } from 'react';

interface CreateBlockOption<T extends IBlockData>
  extends Omit<IBlock<T>, 'transform'> {}

export function createCustomBlock<T extends IBlockData>(
  block: CreateBlockOption<T> & {
    render: (
      data: T,
      idx: string | null,
      mode: 'testing' | 'production',
      context?: IPage,
      dataSource?: { [key: string]: any }
    ) => IBlockData | ReactElement | null;
  }
): IBlock<T> {
  return {
    ...block,
  };
}
