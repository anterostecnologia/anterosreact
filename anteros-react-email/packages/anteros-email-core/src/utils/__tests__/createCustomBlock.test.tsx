import { BlockManager } from '../BlockManager';
import { BasicType } from '@core/constants';
import { createCustomBlock } from '../createCustomBlock';
import { merge } from 'lodash';
import { Section, Column, Text, Image, Button } from '../../components';
import { IBlockData } from '@core/typings';
import React from 'react';
import { JsonToMjml } from '..';

type IMyFirstBlock = IBlockData<
  {
    'background-color': string;
    'text-color': string;
  },
  {
    buttonText: string;
    imageUrl: string;
  }
>;

const myFirstBlock = createCustomBlock({
  name: 'My first block',
  type: 'MY_FIRST_BLOCK',
  create(payload) {
    const defaultData: IMyFirstBlock = {
      type: 'MY_FIRST_BLOCK',
      data: {
        value: {
          buttonText: 'Got it',
          imageUrl:
            'https://infra.relevantsolutions.com.br/nextcloud/index.php/s/xwyw3TYqA7XSTD4/download',
        },
      },
      attributes: {
        'background-color': '#4A90E2',
        'text-color': '#ffffff',
      },
      children: [],
    };
    return merge(defaultData, payload);
  },
  validParentType: [BasicType.PAGE, BasicType.WRAPPER],
  render(data: IMyFirstBlock) {
    const { imageUrl, buttonText } = data.data.value;
    const attributes = data.attributes;

    const instance = (
      <Section padding='20px'>
        <Column>
          <Image padding='0px 0px 0px 0px' width='100px' src={imageUrl} />
          <Button
            background-color={attributes['background-color']}
            color={attributes['text-color']}
            href='#'
          >
            {buttonText}
          </Button>
        </Column>
      </Section>
    );
    return instance;
  },
});

describe('Test createCustomBlock', () => {
  BlockManager.registerBlocks({ myFirstBlock });

  const pageBlock = BlockManager.getBlockByType(BasicType.PAGE)!;

  it('should render as expected', () => {
    expect(
      JsonToMjml({
        data: pageBlock.create({
          children: [myFirstBlock.create()],
        }),
        mode: 'production',
      })
    ).toMatchSnapshot();
  });
});
