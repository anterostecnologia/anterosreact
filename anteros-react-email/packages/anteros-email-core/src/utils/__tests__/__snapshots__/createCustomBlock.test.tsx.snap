// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test createCustomBlock should render as expected 1`] = `
"
        <mjml>
          <mj-head>

    <mj-html-attributes>
      <mj-html-attribute class=\\"anteros-email\\" multiple-attributes=\\"false\\" attribute-name=\\"text-color\\" text-color=\\"#000000\\"></mj-html-attribute>
<mj-html-attribute class=\\"anteros-email\\" multiple-attributes=\\"false\\" attribute-name=\\"font-family\\" font-family=\\"lucida Grande,Verdana,Microsoft YaHei\\"></mj-html-attribute>
<mj-html-attribute class=\\"anteros-email\\" multiple-attributes=\\"false\\" attribute-name=\\"font-size\\" font-size=\\"14px\\"></mj-html-attribute>
<mj-html-attribute class=\\"anteros-email\\" multiple-attributes=\\"false\\" attribute-name=\\"line-height\\" line-height=\\"1.7\\"></mj-html-attribute>
<mj-html-attribute class=\\"anteros-email\\" multiple-attributes=\\"false\\" attribute-name=\\"responsive\\" responsive=\\"true\\"></mj-html-attribute>

    </mj-html-attributes>




              <mj-breakpoint width=\\"480px\\" />
            <mj-attributes>

              <mj-all font-family=\\"lucida Grande,Verdana,Microsoft YaHei\\" />
              <mj-text font-size=\\"14px\\" />
              <mj-text color=\\"#000000\\" />
              <mj-text line-height=\\"1.7\\" />


            </mj-attributes>
          </mj-head>
          <mj-body background-color=\\"#efeeea\\" width=\\"600px\\" css-class=\\"mjml-body\\">

              <mj-section padding=\\"20px\\" background-repeat=\\"repeat\\" background-size=\\"auto\\" background-position=\\"top center\\" border=\\"none\\" direction=\\"ltr\\" text-align=\\"center\\">

              <mj-column padding=\\"0px 0px 0px 0px\\" border=\\"none\\" vertical-align=\\"top\\">
               <mj-image align=\\"center\\" height=\\"auto\\" padding=\\"0px 0px 0px 0px\\" src=\\"https://assets.maocanhua.cn/10dada65-c4fb-4b1f-837e-59a1005bbea6-image.png\\" width=\\"100px\\"></mj-image>

          <mj-button align=\\"center\\" background-color=\\"#4A90E2\\" color=\\"#ffffff\\" font-weight=\\"normal\\" border-radius=\\"3px\\" padding=\\"10px 25px 10px 25px\\" inner-padding=\\"10px 25px 10px 25px\\" line-height=\\"120%\\" target=\\"_blank\\" vertical-align=\\"middle\\" border=\\"none\\" text-align=\\"center\\" href=\\"#\\">
           Got it
          </mj-button>

              </mj-column>

              </mj-section>

          </mj-body>
        </mjml>
        "
`;
